### ISSUE TITLE
Bare-bones fork?

### ISSUE THREAD

**Issue by SaFrMo** [funkhaus/vuepress#82]
In writing documentation for Vuepress, I'm seeing how opinionated the framework is - we have solid methods for combining Vue and WordPress, but those methods require a lot of explanation and documentation.

What are y'all's thoughts on my creating a bare-bones fork or branch that serves as a more public-facing framework, then maintaining the current version as the in-house version of Vuepress? 

I think it's important to make the learning curve as gentle as possible for new users to encourage them to learn and, hopefully, contribute, and paring down VP to the bare minimum could open up the door for more useful components from a wider audience of devs. Combining Vue and WordPress is a problem that developers are definitely interested in solving, I just don't want VP to be so specialized to Funkhaus that it alienates people who could use and build the ecosystem of this tool.

@drewbaker @jrobson153 @lwaldner thoughts?

**Comment by drewbaker** [comment]
I thought we were doing an OK job at keeping this generic enough to be useful. Curious what you think would be better?

**Comment by jrobson153** [comment]
I definitely do agree that it's a tad specific to our needs. It's tough because unlike Rest Easy which is a simple tool that just does one specific thing, Vuepress is like a mash of a bunch of different things and has a lot of dependencies. So trying to document it I imagine is difficult because it involves explaining a lot of other stuff.

The bare bones fork is in interesting idea, but I would just worry that it would get too messy trying to maintain both. There would probably be a lot of work that would need to be duplicated between them whenever we change our process on something.

Maybe the best move is to just commit to either really being critical about only including features that are useful generically vs features that are only useful to Funkhaus, or just decide we don't care and keep building it out as our own internal framework.

**Comment by drewbaker** [comment]
Would like to hear some examples of specific things that we are doing that are Funkhaus specific, but I wonder if there is a way to do something like `fh-components` but for Vuepress. I guess that is kind of like a child theme? So you have the generic Vuepress, and then you add on this layer of Funkhaus specific tools/templates/etc? That would then allow other people to develop there only toolset on top of it as well.

**Comment by jrobson153** [comment]
I think the "opinionated" parts come from a lot of different places. You _must_ use Rest Easy and understand it, you _must_ use vuex and the vue router and conform to the Rest Easy structure of loop/site/meta. The hamburger, clickblocker, and menu structure being included in the theme maybe only makes sense to us. Most of the commented out code examples are centered around "Director - Spot 1" type naming that only applies to us.

Then there's just a lot of bias built around how we use wordpress specifically, assuming that most of the structure of the site will come from the hierarchy of pages when a lot of other WP developers will lean more on custom post types, taxonomies, even users and roles, etc.

It's all little stuff that I think makes sense because of the way we use it, but all adds up to a very steep learning curve for anyone coming at it that's not familiar with the Funkhaus school of thought or is building a different type of website or app. 

Basically, IMO if all you're trying to do is just have a decent way of putting Wordpress and Vue together, this framework will probably be more confusing than helpful in its current state.

**Comment by SaFrMo** [comment]
John nailed the problem I'm running into with documentation:

> Vuepress is like a mash of a bunch of different things and has a lot of dependencies. So trying to document it I imagine is difficult because it involves explaining a lot of other stuff.

For example:

* Like John mentioned, the boilerplate components (hamburger, clickblocker, slideshow template) are super handy for Funkhaus sites, but not necessarily for a generic WordPress theme.
* Ditto for the fh-components commented out in `main.js` and the examples in `functions/router.php` - really useful for FH sites but confusing to new users (and potentially irritating to have to go in and delete every time for experienced VP users).
* Lots of assumptions being made in the tracking, og-tags, and schema files in `parts/` - I'd argue that these are 99% good assumptions, but if a user didn't want Google analytics, that's another file to delete on each install.

I'd love to see a version of Vuepress that looked like:

```
functions
-- recommended plugins
-- custom templates list
-- image sizes
-- rest-easy filters
-- router
-- wp-functions
parts
-- font-loader
-- og-tags
-- schema
src
-- utils
-- -- cache-crawler
-- -- router
-- -- store
-- views
-- -- Default.vue
static
-- (contents here can be the same as we have them now)
(misc config files, readme, license, index.php, style.css)
```

And had separate VP plugins for:
* Developer role
* Shortcode to Vue component translation
* GA tracking

And NPM packages for:
* svg-image (if possible? I know it can't be an fh-component so this might not work)
* contents of `src/styles` folder

This would definitely be extra work to maintain, but I think the ability to select a Vue component from a drop-down list in the WP backend, then see that site in action as a SPA with smooth transitions and fast loading times, could be a great selling point. From there, it'll be easier to explain to developers why Funkhaus makes the choices it does.

**Comment by SaFrMo** [comment]
Turns out making the stripped-down version is really useful for working on the documentation anyway, so I'm going to push a branch called `basic` with this structure in place. Even if we never end up maintaining it, it'll be really useful for illustrative purposes in the docs/tutorial. 

I guess this issue should probably be more about how much effort we want to put into maintaining the `basic` branch? Seems like a thing we should consider!

**Comment by jrobson153** [comment]
@SaFrMo I agree about the structure you're proposing for the stripped down version, and definitely think this is something we need to do if we ever hope to get other people adopting Vuepress. For now though I think the amount of effort we want to put into this is not very much.

Feel free to chip away at making the basic branch if it helps the documentation you've been working on, but let's avoid putting any significant effort into building or maintaining a stripped down version of it until we're ready to take that on.

**Comment by SaFrMo** [comment]
That answers my question! We can keep talking about this in the future, but I'll close this issue for now. Thanks, all!

**Comment by drewbaker** [comment]
If I could give you some notes on what I struggled with when learning Vue and Vuepress.

I'd change the first half of the readme to be a "Getting started guide" rather than just explanation of technical details like it is now.

Basically, you should be able to follow along and get a simple example of WordPress like you would with the "2018 Theme" out of the box. A home page that loops blog posts, some about pages etc.

1. Start with an explanation of how Vue needs JSON data out of WordPress, and Rest Easy is a complementary plugin made by us that you need to do that. Really just needs to be one paragraph.
2. Simple example showing how to display the loop. Include examples of Vue code and command line instructions you need. This would lead in to a simple demo of how to setup Router.
3. If you want to display the current page, here is how. Also show how to setup Router for this, and how dev_id's help with this.
4. How to use related and meta in Vue.
5. Why and how to write a simple rest easy filter to handle something like a custom post type. 
6. When to use Getters, Lodash _get() etc..
7. How to use SVGs and a link to fh-components for responsive images and wp-content.
8. Our recommended file structure
9. Deploying
10. Links to useful things, like examples, Vue, Lodash etc.
11. Urge people to ask us questions and open issues.

**The most confusing thing for me was Rest Easy filters.** Just when things get serialized and how to avoid infinite loops is not very intuitive. I wish there was a way where we kept all that stuff as regular PHP arrays/objects, and it was parsed into JSON before being sent to the frontend. Some stuff being an array, and some stuff being objects is really confusing. Adding more context (like what post it is operating on etc...) to each filter would probably help a lot too.

**Comment by SaFrMo** [comment]
This is fantastic info, thanks so much @drewbaker ! I'll definitely incorporate it into the tutorial and docs.
